core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1018,34 +1033,48 @@ private IsReadyForExecutionResult isReadyForExecution(SubPlan subPlan)
                     return IsReadyForExecutionResult.notReady();
                 }
 
-                Optional<OutputDataSizeEstimateResult> result = sourceStageExecution.getOutputDataSize(stageExecutions::get);
+                Optional<OutputDataSizeEstimateResult> result = sourceStageExecution.getOutputDataSize(stageExecutions::get, eager);
                 if (result.isEmpty()) {
                     return IsReadyForExecutionResult.notReady();
                 }
 
                 switch (result.orElseThrow().getStatus()) {
                     case ESTIMATED_BY_PROGRESS -> estimatedByProgressSourcesCount++;
                     case ESTIMATED_BY_SMALL_INPUT -> estimatedBySmallInputSourcesCount++;
+                    case ESTIMATED_FOR_EAGER_PARENT -> estimatedForEagerParent++;
                     default -> throw new IllegalStateException(format("unexpected status %s", result.orElseThrow().getStatus())); // FINISHED handled above
                 }
 
                 sourceOutputSizeEstimates.put(sourceStageExecution.getStageId(), result.orElseThrow().getOutputDataSizeEstimate());
                 someSourcesMadeProgress = someSourcesMadeProgress || sourceStageExecution.isSomeProgressMade();
             }
 
-            if (!subPlan.getChildren().isEmpty() && !someSourcesMadeProgress) {
+            if (!subPlan.getChildren().isEmpty() && !someSourcesMadeProgress && !eager) {
                 return IsReadyForExecutionResult.notReady();
             }
 
             if (speculative) {
-                log.debug("scheduling speculative %s/%s; sources: finished=%s; estimatedByProgress=%s; estimatedSmall=%s",
+                log.debug("scheduling speculative %s/%s; sources: finished=%s; estimatedByProgress=%s; estimatedSmall=%s; estimatedForEagerParent=%s",
                         queryStateMachine.getQueryId(),
                         subPlan.getFragment().getId(),
                         finishedSourcesCount,
                         estimatedByProgressSourcesCount,
-                        estimatedBySmallInputSourcesCount);
+                        estimatedBySmallInputSourcesCount,
+                        estimatedForEagerParent);
             }
-            return IsReadyForExecutionResult.ready(sourceOutputSizeEstimates.buildOrThrow(), false);
+            return IsReadyForExecutionResult.ready(sourceOutputSizeEstimates.buildOrThrow(), eager);
+        }
+
+        private boolean shouldScheduleEagerly(SubPlan subPlan)
+        {
+            return hasSmallFinalLimitNode(subPlan);
+        }
+
+        private static boolean hasSmallFinalLimitNode(SubPlan subPlan)
+        {
+            return PlanNodeSearcher.searchFrom(subPlan.getFragment().getRoot())
+                    .where(node -> node instanceof LimitNode limitNode && !limitNode.isPartial() && limitNode.getCount() < 1_000_000)
```

Comment:
For my knowledge, what does it mean if a limit node is "partial"?

Reply:
non-final. Limit is planned as

```
       LIMIT_FINAL (single)
                 |
       LIMIT_PARTIAL(distributed)
   ```
There is just a single task executing `LIMIT_FINAL` while there is many of `LIMIT_PARTIAL` nodes. Each `LIMIT_PARTIAL` would output up-to limit number of rows, and `LIMIT_FINAL` will do final selection